swupd/manifest.go

```diff
@@ -386,59 +386,100 @@ func (m *Manifest) prependFile(f *File) {
 // At this point Manifest m should only have the files that were present
 // in the chroot for that manifest. Link delta peers with the oldManifest
 // if the file in the oldManifest is not deleted or ghosted
-func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) (int, int, int) {
+	m.sortFilesName()
+	oldManifest.sortFilesName()
+
 	changed := 0
-	for _, of := range oldManifest.Files {
-		if match := of.findFileNameInSlice(m.Files); match != nil {
+	added := 0
+	deleted := 0
+	i := 0
+	j := 0
+	for i < len(m.Files) && j < len(oldManifest.Files) {
+		nf := m.Files[i]
+		of := oldManifest.Files[j]
+		if nf.Name == of.Name {
+			// if it is the same name check if anything about the file has changed.
+			// if something has changed update the version to current version and
+			// record that the file has changed
 			if of.Status == statusDeleted || of.Status == statusGhosted {
+				// don't worry about ghosted or deleted files
+				i++
+				j++
 				continue
 			}
 
-			match.DeltaPeer = of
-			of.DeltaPeer = match
-			if !sameFile(match, of) {
-				match.Version = m.Header.Version
+			// set up peers
+			nf.DeltaPeer = of
+			of.DeltaPeer = nf
+			if !sameFile(nf, of) {
+				// if the file isn't exactly the same, record the change
+				// and update the version
+				nf.Version = m.Header.Version
 				changed++
 			} else {
-				match.Version = of.Version
+				// otherwise the version is the same as the old file
+				nf.Version = of.Version
+			}
+			// advance indices for both file lists since we had a match
+			i++
+			j++
+		} else if nf.Name < of.Name {
+			// if the file does not exist in the old manifest it is a new
+			// file in this manifest. Update the version and record that
+			// it is added.
+			nf.Version = m.Header.Version
+			added++
+			// look at next file in current manifest
+			i++
+		} else {
+			// if the file exists in the old manifest and does not *yet* exist
+			// in the new manifest, it was deleted. Set up the deleted file
+			// and prepend it to the files list via newDeleted()
+			if of.Status == statusDeleted || of.Status == statusGhosted {
```

Comment:
I am finding I am writing this quite a bit as well. Perhaps we might think about splitting this into two values, like the old is_deleted and is_ghosted bits?

Reply:
Yes that might be worth doing. It is kind of weird because with the current manifest format these flags occupy the same position (`.d/g..`) which is why they were combined into their own `enum`.

What if we instead check if `file.Status == statusUnset`? Not very extensible if we add more status flags, but much simpler.

Reply:
What about having a method in fstatus, so we can say things like: of.Status.IsGone()? (name to be defined)

Reply:
@cmarcelo that sounds like a good idea. I'd rather leave that for another PR though.